cloudevents/work/agent/client/manifestwork.go

```diff
@@ -26,117 +27,123 @@ const (
 
-	_, err := c.lister.Get(manifestWork.Name)
+func (c *ManifestWorkAgentClient) Patch(ctx context.Context, name string, pt kubetypes.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (result *workv1.ManifestWork, err error) {
+	klog.V(4).Infof("patching manifestwork %s", name)
+
+	lastWork, err := c.lister.Get(name)
 	if err != nil {
 		return nil, err
 	}
 
-	eventDataType, err := types.ParseCloudEventsDataType(manifestWork.Annotations[codec.CloudEventsDataTypeAnnotationKey])
+	patchedWork, err := utils.Patch(pt, lastWork, data)
 	if err != nil {
 		return nil, err
 	}
 
-	updatedWork := manifestWork.DeepCopy()
+	eventDataType, err := types.ParseCloudEventsDataType(patchedWork.Annotations[codec.CloudEventsDataTypeAnnotationKey])
+	if err != nil {
+		return nil, err
+	}
 
 	eventType := types.CloudEventsType{
 		CloudEventsDataType: *eventDataType,
 		SubResource:         types.SubResourceStatus,
-		Action:              UpdateRequestAction,
 	}
 
-	if err := c.cloudEventsClient.Publish(ctx, eventType, updatedWork); err != nil {
-		return nil, err
-	}
-	return updatedWork, nil
-}
+	newWork := patchedWork.DeepCopy()
 
-func (c *ManifestWorksAgentClient) Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error {
-	klog.Fatal("Delete function for ManifestWorksAgentClient is unsupported")
-	return nil
-}
+	if !newWork.DeletionTimestamp.IsZero() && len(newWork.Finalizers) == 0 {
+		// the finalizers of a deleting manifestwork are removed, marking the manifestwork status to deleted and sending
+		// it back to source
+		meta.SetStatusCondition(&newWork.Status.Conditions, metav1.Condition{
+			Type:    ManifestsDeleted,
+			Status:  metav1.ConditionTrue,
+			Reason:  "ManifestsDeleted",
+			Message: fmt.Sprintf("The manifests are deleted from the cluster %s", newWork.Namespace),
+		})
 
-func (c *ManifestWorksAgentClient) DeleteCollection(ctx context.Context, opts metav1.DeleteOptions, listOpts metav1.ListOptions) error {
-	klog.Fatal("DeleteCollection function for ManifestWorksAgentClient is unsupported")
-	return nil
-}
+		eventType.Action = DeleteRequestAction
+		if err := c.cloudEventsClient.Publish(ctx, eventType, newWork); err != nil {
+			return nil, err
+		}
 
-func (c *ManifestWorksAgentClient) Get(ctx context.Context, name string, opts metav1.GetOptions) (*workv1.ManifestWork, error) {
-	klog.V(4).Infof("getting manifestwork %s", name)
-	return c.lister.Get(name)
-}
+		// delete the manifestwork from the ManifestWorkInformer local cache.
+		c.watcher.Receive(watch.Event{Type: watch.Deleted, Object: newWork})
+		return newWork, nil
+	}
 
-func (c *ManifestWorksAgentClient) List(ctx context.Context, opts metav1.ListOptions) (*workv1.ManifestWorkList, error) {
-	klog.V(4).Infof("sync manifestworks")
-	// send resync request to fetch manifestworks from source when the ManifestWorkInformer status
-	if err := c.cloudEventsClient.Resync(ctx); err != nil {
-		return nil, err
+	if IsStatusUpdate(subresources) {
```

Comment:
if it is not status subresource we should return error

Reply:
return the error at the end

Reply:
return the error if the patch is not for status or finalizers and allow update finalizers(add/remvoe)